src/Ormolu/Printer/Meat/Pragma.hs

```diff
@@ -8,18 +8,43 @@ module Ormolu.Printer.Meat.Pragma
   )
 where
 
+import Data.List (isPrefixOf)
 import qualified Data.Set as S
 import qualified Data.Text as T
 import Ormolu.Parser.Pragma (Pragma (..))
 import Ormolu.Printer.Combinators
 
-data PragmaTy = Language | OptionsGHC | OptionsHaddock
+-- | Pragma classification.
+data PragmaTy
+  = Language LanguagePragmaClass
+  | OptionsGHC
+  | OptionsHaddock
+  deriving (Eq, Ord)
+
+-- | Language pragma classification.
+--
+-- The order in which language pragmas are put in the input sometimes
+-- matters. This is because some language extensions can enable other
+-- extensions, yet the extensions coming later in the list have the ability
+-- to change it. So here we classify all extensions by assigning one of the
+-- four groups to them. Then we only sort inside of the groups.
```

Comment:
It would be nice to specify that the Ord instance specifies the order of the pragmas.

It would also be nice to link to the issue tracker, because singling out CUSKs and ImplicitPrelude is definitely a head-scratcher, even despite the comment above.